File: app/src/test/java/com/twilio/video/app/TestAuthModule.java
package com.twilio.video.app;

import android.app.Application;
import android.content.SharedPreferences;

import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.twilio.video.app.auth.AuthenticationProvider;
import com.twilio.video.app.auth.Authenticator;
import com.twilio.video.app.auth.EmailAuthenticator;
import com.twilio.video.app.auth.FirebaseAuthenticator;
import com.twilio.video.app.auth.FirebaseWrapper;
import com.twilio.video.app.auth.GoogleAuthProviderWrapper;
import com.twilio.video.app.auth.GoogleAuthWrapper;
import com.twilio.video.app.auth.GoogleAuthenticator;
import com.twilio.video.app.auth.GoogleSignInOptionsBuilderWrapper;
import com.twilio.video.app.auth.GoogleSignInWrapper;

import java.util.ArrayList;
import java.util.List;

import dagger.Module;
import dagger.Provides;

@Module(includes = {
        ApplicationModule.class,
        TestWrapperAuthModule.class
}
)
public class TestAuthModule {

    @Provides
    @ApplicationScope
    Authenticator providesAuthenticator(
            FirebaseWrapper firebaseWrapper,
            GoogleAuthWrapper googleAuthWrapper,
            GoogleSignInWrapper googleSignInWrapper,
            GoogleSignInOptionsBuilderWrapper googleSignInOptionsBuilderWrapper,
            GoogleAuthProviderWrapper googleAuthProviderWrapper,
            Application application,
            SharedPreferences sharedPreferences) {
        List<AuthenticationProvider> authenticators = new ArrayList<>();
        authenticators.add(
                new GoogleAuthenticator(
                        firebaseWrapper,
                        application,
                        googleAuthWrapper,
                        googleSignInWrapper,
                        googleSignInOptionsBuilderWrapper,
                        googleAuthProviderWrapper,
                        sharedPreferences)
        );
        authenticators.add(new EmailAuthenticator(firebaseWrapper, sharedPreferences));
        return new FirebaseAuthenticator(firebaseWrapper, authenticators);
    }
}

File: app/src/test/java/com/twilio/video/app/TestAuthModule.kt
package com.twilio.video.app

import android.app.Application
import com.twilio.video.app.auth.AuthenticationProvider
import com.twilio.video.app.auth.Authenticator
import com.twilio.video.app.auth.EmailAuthProvider
import com.twilio.video.app.auth.FirebaseAuthenticator
import com.twilio.video.app.auth.FirebaseWrapper
import com.twilio.video.app.auth.GoogleAuthProvider
import com.twilio.video.app.auth.GoogleAuthProviderWrapper
import com.twilio.video.app.auth.GoogleAuthWrapper
import com.twilio.video.app.auth.GoogleSignInOptionsWrapper
import com.twilio.video.app.auth.GoogleSignInWrapper
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent

@Module
@InstallIn(SingletonComponent::class)
class TestAuthModule {

    @Provides
    internal fun providesAuthenticator(
        firebaseWrapper: FirebaseWrapper,
        googleAuthWrapper: GoogleAuthWrapper,
        googleSignInWrapper: GoogleSignInWrapper,
        googleSignInOptionsWrapper: GoogleSignInOptionsWrapper,
        googleAuthProviderWrapper: GoogleAuthProviderWrapper,
        application: Application
    ): Authenticator {
        val authenticators = ArrayList<AuthenticationProvider>()
        authenticators.add(
                GoogleAuthProvider(
                        firebaseWrapper,
                        application,
                        googleAuthWrapper,
                        googleSignInWrapper,
                        googleSignInOptionsWrapper,
                        googleAuthProviderWrapper)
        )
        authenticators.add(EmailAuthProvider(firebaseWrapper))
        return FirebaseAuthenticator(firebaseWrapper, authenticators)
    }
}
